x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlIndexAndAlias.java

```diff
@@ -215,4 +219,52 @@ private static void updateWriteAlias(Client client,
                 listener::onFailure),
             client.admin().indices()::aliases);
     }
+
+    /**
+     * Installs the index template specified by {@code templateConfig} if it is not in already
+     * installed in {@code clusterState}.
+     *
+     * The check for presence is simple and will return the listener on
+     * the calling thread if successful. If the template has to be installed
+     * an async call will be made.
+     *
+     * @param clusterState The cluster state
+     * @param client For putting the template
+     * @param templateConfig The config
+     * @param listener Async listener
+     */
+    public static void installIndexTemplateIfRequired(
+        ClusterState clusterState,
+        Client client,
+        IndexTemplateConfig templateConfig,
+        ActionListener<Boolean> listener
+    ) {
+        String templateName = templateConfig.getTemplateName();
+
+        // The check for existence of the template is against the cluster state, so very cheap
+        if (haveIndexTemplate(clusterState, templateName)) {
+            listener.onResponse(true);
+            return;
+        }
+
+        PutIndexTemplateRequest request = new PutIndexTemplateRequest(templateName)
+            .source(templateConfig.loadBytes(), XContentType.JSON);
+        request.masterNodeTimeout(TimeValue.timeValueMinutes(1));
+
+        ActionListener<AcknowledgedResponse> innerListener = ActionListener.wrap(
+            response ->  {
+                if (response.isAcknowledged() == false) {
+                    logger.error("error adding legacy template [{}], request was not acknowledged", templateName);
```

Comment:
Should we make this a `warn`? If it's an error I would expect it causes `listener.onFailure` to fire.

Reply:
It is consistent with `IndexTemplateRegistry` 

https://github.com/elastic/elasticsearch/blob/3b68df23559c19fa4823c9e661ed5b25c3281fab/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java#L303


TBH I wasn't sure what it would mean if `isAcknowledged == false` but I agree it makes more sense as a warning so I've changed it to warning. 

Looking at the put template code the response cannot be false but that might change

https://github.com/elastic/elasticsearch/blob/261250d616db6e167a7dd7ab45f88e1d4daf210e/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java#L763